File: src/task.rs
//! Task discovery: shell out to `task --list-all` and parse the result.
//!
//! The obvious call is `--list-all --json`, which additionally carries aliases as an
//! array and each task's source location. It is also, measured on a repo with seven
//! `includes:` and twenty tasks declaring `sources:`, **fifty-six times slower** — 4.28s
//! against 0.076s. The `--json` form computes `up_to_date` for every task, which means
//! fingerprinting every `sources:` glob, which on a Rust workspace means walking
//! `target/`. Almost all of that 4.28s was system time, not parsing.
//!
//! Four seconds before the first frame is not a price worth paying for a source location
//! nothing reads yet, so this parses the text form instead. If the file pivot or
//! jump-to-definition ever land, they should fetch the JSON on a background thread and
//! fill it in — not block startup on it.
//!
//! Neither form carries dependencies, so this is only the static half of the picture; the
//! execution graph comes from [`crate::graph`].

use anyhow::{bail, Context, Result};
use std::path::Path;
use std::process::Command;

#[derive(Debug, Clone)]
pub struct Task {
    /// Full colon path, e.g. `backend:migrate:down`.
    pub name: String,
    pub desc: String,
    pub aliases: Vec<String>,
    /// Heuristic: does the description suggest this touches production or destroys data?
    pub dangerous: bool,
}

impl Task {
    /// Path segments of the name: `backend:migrate:down` -> ["backend", "migrate", "down"].
    pub fn segments(&self) -> Vec<&str> {
        self.name.split(':').collect()
    }

    /// The last segment — the "verb" the task pivot groups on.
    pub fn verb(&self) -> &str {
        self.name.rsplit(':').next().unwrap_or(&self.name)
    }

    /// Usage hint mined from the description, which by convention spells out an example:
    /// "Scaffold a migration and register it (NAME=add_x)" or "task backend:test -- -p
    /// ingest for one crate".
    ///
    /// Shown beside the args prompt rather than pre-filled into it. The descriptions trail
    /// off into prose often enough — "…-- -p ingest for one crate" — that pre-filling
    /// would hand you a command that is wrong in a way you might not notice before
    /// pressing enter.
    pub fn args_hint(&self) -> Option<String> {
        // Descriptions are written from inside the file that defines the task, so an
        // included one often spells its example with the bare name: `site/Taskfile.yml`
        // says "task new -- …" for what the root calls `site:new`. Try both.
        let mut needles = vec![format!("task {} ", self.name)];
        if self.verb() != self.name {
            needles.push(format!("task {} ", self.verb()));
        }
        for needle in &needles {
            let Some(at) = self.desc.find(needle.as_str()) else {
                continue;
            };
            let hint = trim_prose(&self.desc[at + needle.len()..]);
            if !hint.is_empty() {
                return Some(hint);
            }
        }
        // Bare `NAME=value` conventions, e.g. "(NAME=add_x)" or "(WORD=адрес)".
        let open = self.desc.find('(')?;
        let close = self.desc[open..].find(')')? + open;
        let inner = self.desc[open + 1..close].trim();
        let looks_like_assignment = inner.split_whitespace().next().is_some_and(|w| {
            w.contains('=') && w.chars().next().is_some_and(|c| c.is_ascii_uppercase())
        });
        looks_like_assignment.then(|| inner.to_string())
    }
}

/// The `KEY=` prefixes implied by a hint like `NAME=backend` or `WORD=адрес`.
///
/// A fallback for tasks that take a variable but do not declare `requires:`. Only the key
/// is kept — pre-filling the example *value* would be handing you someone else's argument.
pub fn keys_in_hint(hint: &str) -> Vec<String> {
    hint.split_whitespace()
        .filter_map(|w| w.split_once('='))
        .map(|(k, _)| k)
        .filter(|k| {
            !k.is_empty()
                && k.chars()
                    .all(|c| c.is_ascii_uppercase() || c == '_' || c.is_ascii_digit())
        })
        .map(|k| k.to_string())
        .collect()
}

/// Cut a mined hint back to the part that is plausibly an argument.
///
/// Descriptions run on: "-- infra:deploy:plan`.", "NAME=backend (the BRANCH survives)".
/// The parenthetical and the trailing punctuation are commentary, not arguments.
fn trim_prose(rest: &str) -> String {
    let mut end = rest.len();
    for stop in [")", " (", ";", ", "] {
        if let Some(at) = rest.find(stop) {
            end = end.min(at);
        }
    }
    rest[..end]
        .trim()
        .trim_end_matches(['.', ',', '`', ' '])
        .to_string()
}

/// Split an args line the way a shell would, so quoted arguments survive.
///
/// `site:new -- "My Post Title"` has to reach go-task as one argument, not three.
pub fn split_args(input: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    let mut started = false;
    let mut chars = input.chars().peekable();

    while let Some(c) = chars.next() {
        match c {
            '\\' => {
                if let Some(next) = chars.next() {
                    current.push(next);
                    started = true;
                }
            }
            '\'' | '"' => match quote {
                Some(q) if q == c => quote = None,
                Some(_) => {
                    current.push(c);
                    started = true;
                }
                // An empty quoted string is still an argument.
                None => {
                    quote = Some(c);
                    started = true;
                }
            },
            c if c.is_whitespace() && quote.is_none() => {
                if started {
                    out.push(std::mem::take(&mut current));
                    started = false;
                }
            }
            c => {
                current.push(c);
                started = true;
            }
        }
    }
    if started {
        out.push(current);
    }
    out
}

/// Name of the opt-in file listing tasks that must not be run by accident.
pub const DANGER_FILE: &str = ".taskui-danger";

/// Match a task name against a pattern supporting `*`.
///
/// Deliberately not a full glob: task names are colon paths, and `deploy:*` plus exact
/// names covers every real case without pulling in a dependency whose semantics would
/// then need explaining.
pub fn glob_match(pattern: &str, name: &str) -> bool {
    let mut parts = pattern.split('*');
    let Some(first) = parts.next() else {
        return false;
    };
    if !name.starts_with(first) {
        return false;
    }
    let mut rest = &name[first.len()..];
    let mut last: Option<&str> = None;
    for part in parts {
        last = Some(part);
        if part.is_empty() {
            continue;
        }
        match rest.find(part) {
            Some(at) => rest = &rest[at + part.len()..],
            None => return false,
        }
    }
    // A trailing `*` swallows whatever is left; otherwise the pattern must reach the end.
    match last {
        None => rest.is_empty(),
        Some("") => true,
        Some(tail) => name.ends_with(tail),
    }
}

/// Read `.taskui-danger`: one pattern per line, `#` comments, blanks ignored.
///
/// Its presence switches off the description heuristic entirely. A guess and a
/// declaration disagreeing about which tasks are dangerous is worse than either alone —
/// once you have written the list down, that list is the answer.
pub fn danger_patterns(dir: &Path) -> Vec<String> {
    let Ok(text) = std::fs::read_to_string(dir.join(DANGER_FILE)) else {
        return Vec::new();
    };
    text.lines()
        .map(|l| l.split('#').next().unwrap_or("").trim())
        .filter(|l| !l.is_empty())
        .map(|l| l.to_string())
        .collect()
}

/// Words that mark a task as one you should not fire off from a fuzzy filter by accident.
///
/// This is a heuristic over descriptions and it will both over- and under-match. It is a
/// stopgap: the real fix is an explicit marker in the Taskfile, which we can honour later
/// without changing anything else here.
const DANGER_WORDS: [&str; 6] = [
    "production",
    "prod database",
    "applies!",
    "wipe",
    "destroy",
    "claims ",
];

fn looks_dangerous(name: &str, desc: &str) -> bool {
    let d = desc.to_lowercase();
    if DANGER_WORDS.iter().any(|w| d.contains(w)) {
        return true;
    }
    // `backend:migrate:prod`, `backend:promo:prod`, `deploy:*`
    name.ends_with(":prod") || name.starts_with("deploy:") || name == "deploy"
}

/// One line of `task --list-all`:
///
/// ```text
/// * build:      Build all components                    (aliases: b)
/// ```
fn parse_entry(line: &str) -> Option<Task> {
    let rest = line.strip_prefix("* ")?;
    // Names contain colons, so split at the first whitespace rather than the first colon
    // — `backend:migrate:down:` is one name, not three.
    let (name, tail) = match rest.find(char::is_whitespace) {
        Some(at) => (&rest[..at], &rest[at..]),
        None => (rest, ""),
    };
    let name = name.trim_end_matches(':');
    if name.is_empty() {
        return None;
    }

    let mut desc = tail.trim().to_string();
    let mut aliases = Vec::new();
    // Parsed off the end rather than searched for, so a description that happens to
    // mention the word does not get eaten.
    if desc.ends_with(')') {
        if let Some(at) = desc.rfind("(aliases: ") {
            aliases = desc[at + "(aliases: ".len()..desc.len() - 1]
                .split(',')
                .map(|a| a.trim().to_string())
                .filter(|a| !a.is_empty())
                .collect();
            desc = desc[..at].trim().to_string();
        }
    }

    Some(Task {
        dangerous: looks_dangerous(name, &desc),
        name: name.to_string(),
        desc,
        aliases,
    })
}

/// Run `task --list-all` in `dir` and return the tasks, minus the `*:default` entries —
/// in a UI where a namespace is itself a selectable row, a task whose only job is "show
/// available tasks" is noise.
pub fn discover(dir: &Path) -> Result<Vec<Task>> {
    let declared = danger_patterns(dir);
    let out = Command::new("task")
        .arg("--list-all")
        .current_dir(dir)
        .output()
        .context("failed to run `task` — is go-task installed and on PATH?")?;

    if !out.status.success() {
        let stderr = String::from_utf8_lossy(&out.stderr);
        bail!(
            "`task --list-all` failed in {}: {}",
            dir.display(),
            stderr.trim()
        );
    }

    let text = String::from_utf8_lossy(&out.stdout);
    let mut tasks: Vec<Task> = text
        .lines()
        .filter_map(parse_entry)
        .filter(|t| t.name != "default" && !t.name.ends_with(":default"))
        .map(|mut t| {
            if !declared.is_empty() {
                t.dangerous = declared.iter().any(|p| glob_match(p, &t.name));
            }
            t
        })
        .collect();

    tasks.sort_by(|a, b| a.name.cmp(&b.name));
    Ok(tasks)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_a_plain_entry() {
        let t =
            parse_entry("* all:                           Everything: format, lint, test, build")
                .unwrap();
        assert_eq!(t.name, "all");
        assert_eq!(t.desc, "Everything: format, lint, test, build");
        assert!(t.aliases.is_empty());
    }

    #[test]
    fn parses_aliases_off_the_end() {
        let t = parse_entry("* build:      Build all components        (aliases: b)").unwrap();
        assert_eq!(t.name, "build");
        assert_eq!(t.desc, "Build all components");
        assert_eq!(t.aliases, ["b"]);
    }

    #[test]
    fn parses_namespaced_names() {
        let t =
            parse_entry("* backend:migrate:down:   Roll back the most recent migration").unwrap();
        assert_eq!(t.name, "backend:migrate:down");
        assert_eq!(t.desc, "Roll back the most recent migration");
    }

    /// A description ending in a parenthesis is not an alias list.
    #[test]
    fn a_parenthesised_description_is_not_eaten() {
        let t = parse_entry("* setup:   Install tools and dependencies (safe to re-run)").unwrap();
        assert_eq!(t.desc, "Install tools and dependencies (safe to re-run)");
        assert!(t.aliases.is_empty());
    }

    /// `--list-all` includes tasks with no description at all.
    #[test]
    fn handles_a_missing_description() {
        let t = parse_entry("* sec:secrets:dir:").unwrap();
        assert_eq!(t.name, "sec:secrets:dir");
        assert_eq!(t.desc, "");
    }

    #[test]
    fn ignores_the_header_and_blank_lines() {
        assert!(parse_entry("task: Available tasks for this project:").is_none());
        assert!(parse_entry("").is_none());
    }

    fn task_with(name: &str, desc: &str) -> Task {
        Task {
            name: name.into(),
            desc: desc.into(),
            aliases: Vec::new(),
            dangerous: false,
        }
    }

    #[test]
    fn mines_a_usage_hint_from_the_description() {
        let t = task_with(
            "backend:test",
            "Run Rust workspace tests (task backend:test -- -p ingest for one crate)",
        );
        assert_eq!(t.args_hint().as_deref(), Some("-- -p ingest for one crate"));
    }

    /// Included tasks describe themselves by their local name.
    #[test]
    fn mines_a_hint_written_with_the_bare_name() {
        let t = task_with(
            "site:new",
            r#"New post — usage: task new -- "My Post Title" (title is slugified)"#,
        );
        assert_eq!(t.args_hint().as_deref(), Some(r#"-- "My Post Title""#));
    }

    /// The commentary after the arguments is not part of them.
    #[test]
    fn trims_trailing_prose_from_a_hint() {
        let t = task_with(
            "wt:rm",
            "Remove an agent's worktree: task wt:rm NAME=backend (the BRANCH survives)",
        );
        assert_eq!(t.args_hint().as_deref(), Some("NAME=backend"));

        let t = task_with(
            "deploy:local",
            "Full deploy. Pass a target with `task deploy:local -- infra:deploy:plan`.",
        );
        assert_eq!(t.args_hint().as_deref(), Some("-- infra:deploy:plan"));

        let t = task_with(
            "dev:run",
            "Run the backend once: task dev:run -- <args>, default `serve`",
        );
        assert_eq!(t.args_hint().as_deref(), Some("-- <args>"));
    }

    #[test]
    fn mines_bare_assignment_conventions() {
        let t = task_with(
            "backend:gen:migration",
            "Scaffold a migration and register it (NAME=add_x)",
        );
        assert_eq!(t.args_hint().as_deref(), Some("NAME=add_x"));
    }

    /// A parenthetical that is just prose is not a usage hint.
    #[test]
    fn does_not_invent_a_hint_from_ordinary_prose() {
        let t = task_with("setup", "Install tools and dependencies (safe to re-run)");
        assert_eq!(t.args_hint(), None);
    }

    /// A hint of the `KEY=value` shape tells us the key even when the task does not
    /// declare `requires:`.
    #[test]
    fn extracts_keys_from_an_assignment_hint() {
        assert_eq!(keys_in_hint("NAME=backend"), ["NAME"]);
        assert_eq!(keys_in_hint("WORD=адрес"), ["WORD"]);
        assert_eq!(keys_in_hint("A=1 B=2"), ["A", "B"]);
    }

    /// `--` style arguments carry no key to pre-fill.
    #[test]
    fn extracts_no_keys_from_a_dash_dash_hint() {
        assert!(keys_in_hint(r#"-- "My Post Title""#).is_empty());
        assert!(keys_in_hint("-- -p ingest").is_empty());
    }

    /// Quoted arguments have to reach go-task intact.
    #[test]
    fn splits_args_like_a_shell() {
        assert_eq!(
            split_args("-- convert report.pdf"),
            ["--", "convert", "report.pdf"]
        );
        assert_eq!(split_args(r#"-- "My Post Title""#), ["--", "My Post Title"]);
        assert_eq!(split_args("NAME=backend"), ["NAME=backend"]);
        assert_eq!(split_args("  spaced   out  "), ["spaced", "out"]);
        assert_eq!(split_args(""), [] as [String; 0]);
    }

    #[test]
    fn quotes_inside_words_and_escapes_survive() {
        assert_eq!(split_args(r#"MSG="hello world""#), ["MSG=hello world"]);
        assert_eq!(split_args(r"path\ with\ spaces"), ["path with spaces"]);
        // An empty quoted string is a real argument, and `--` is one too.
        assert_eq!(split_args(r#"-- '' empty"#), ["--", "", "empty"]);
    }

    #[test]
    fn globs_match_colon_paths() {
        assert!(glob_match("deploy:*", "deploy:backend"));
        assert!(glob_match("deploy:*", "deploy:logs:archive"));
        assert!(!glob_match("deploy:*", "dev:up"));
        assert!(glob_match("backend:migrate:prod", "backend:migrate:prod"));
        assert!(!glob_match("backend:migrate:prod", "backend:migrate:down"));
        assert!(glob_match("*:prod", "backend:promo:prod"));
        assert!(glob_match("*", "anything"));
    }

    /// Production-touching tasks are flagged so they are not one fuzzy keypress away.
    #[test]
    fn flags_production_tasks() {
        assert!(
            parse_entry("* deploy:backend:   Deploy the backend")
                .unwrap()
                .dangerous
        );
        assert!(
            parse_entry("* backend:migrate:prod:   Apply migrations")
                .unwrap()
                .dangerous
        );
        assert!(
            !parse_entry("* backend:lint:   Rust lints")
                .unwrap()
                .dangerous
        );
    }
}

File: internal/task/task.go
// Package task discovers tasks by shelling out to `task --list-all` and parsing the result.
//
// The obvious call is `--list-all --json`, which additionally carries aliases as an array
// and each task's source location. It is also, measured on a repo with seven `includes:`
// and twenty tasks declaring `sources:`, fifty-six times slower — 4.28s against 0.076s.
// The `--json` form computes up_to_date for every task, which means fingerprinting every
// `sources:` glob, which on a large workspace means walking the build directory. Almost
// all of that 4.28s was system time, not parsing.
//
// Four seconds before the first frame is not a price worth paying for a source location
// nothing reads yet, so this parses the text form instead. If the file pivot or
// jump-to-definition ever land, they should fetch the JSON on a background goroutine and
// fill it in — not block startup on it.
//
// Neither form carries dependencies, so this is only the static half of the picture; the
// execution graph comes from the graph package.
package task

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"slices"
	"sort"
	"strings"
	"unicode"

	"github.com/charmbracelet/x/ansi"
)

// Task is one entry of `task --list-all`.
type Task struct {
	// Name is the full colon path, e.g. `backend:migrate:down`.
	Name    string
	Desc    string
	Aliases []string
	// Dangerous is a heuristic: does the description suggest this touches production or
	// destroys data?
	Dangerous bool
	// Where the task is written, once the JSON listing has arrived. Zero until then, which
	// is why the file pivot lists everything ungrouped for the first moment of a session
	// and then fills in.
	Where Where
}

// Segments splits the name: `backend:migrate:down` -> ["backend", "migrate", "down"].
func (t Task) Segments() []string {
	return strings.Split(t.Name, ":")
}

// Verb is the last segment — what the task pivot groups on.
func (t Task) Verb() string {
	if i := strings.LastIndex(t.Name, ":"); i >= 0 {
		return t.Name[i+1:]
	}
	return t.Name
}

// ArgsHint mines a usage hint from the description, which by convention spells out an
// example: "Scaffold a migration and register it (NAME=add_x)" or "task backend:test --
// -p ingest for one crate".
//
// Shown beside the args prompt rather than pre-filled into it. The descriptions trail off
// into prose often enough — "…-- -p ingest for one crate" — that pre-filling would hand
// you a command that is wrong in a way you might not notice before pressing enter.
func (t Task) ArgsHint() (string, bool) {
	hints := t.ArgsHints()
	if len(hints) == 0 {
		return "", false
	}
	return hints[0], true
}

// ArgsHints is every example the description spells out, in the order they appear.
//
// One description often carries more than one — "task backend:test -- -p ingest, or
// task backend:test -- -p api" — and the first is only the one the hint has room for. The
// args prompt completes against the rest, which makes a convention nobody has to adopt
// into a set of presets for Taskfiles nobody is going to edit.
func (t Task) ArgsHints() []string {
	var out []string
	// Descriptions are written from inside the file that defines the task, so an included
	// one often spells its example with the bare name: `site/Taskfile.yml` says
	// "task new -- …" for what the root calls `site:new`. Try both.
	needles := []string{"task " + t.Name + " "}
	if v := t.Verb(); v != t.Name {
		needles = append(needles, "task "+v+" ")
	}
	for _, needle := range needles {
		rest := t.Desc
		for {
			at := strings.Index(rest, needle)
			if at < 0 {
				break
			}
			rest = rest[at+len(needle):]
			if hint := trimProse(rest); hint != "" && !slices.Contains(out, hint) {
				out = append(out, hint)
			}
		}
	}
	if len(out) == 0 {
		if inner, ok := t.parenHint(); ok {
			out = append(out, inner)
		}
	}
	return out
}

// parenHint reads the bare `NAME=value` conventions, e.g. "(NAME=add_x)" or "(WORD=адрес)".
func (t Task) parenHint() (string, bool) {
	open := strings.Index(t.Desc, "(")
	if open < 0 {
		return "", false
	}
	rel := strings.Index(t.Desc[open:], ")")
	if rel < 0 {
		return "", false
	}
	inner := strings.TrimSpace(t.Desc[open+1 : open+rel])
	fields := strings.Fields(inner)
	if len(fields) == 0 {
		return "", false
	}
	w := fields[0]
	if !strings.Contains(w, "=") {
		return "", false
	}
	first := []rune(w)[0]
	if first < 'A' || first > 'Z' {
		return "", false
	}
	return inner, true
}

// KeysInHint returns the `KEY=` prefixes implied by a hint like `NAME=backend` or
// `WORD=адрес`.
//
// A fallback for tasks that take a variable but do not declare `requires:`. Only the key
// is kept — pre-filling the example value would be handing you someone else's argument.
func KeysInHint(hint string) []string {
	out := []string{}
	for w := range strings.FieldsSeq(hint) {
		k, _, ok := strings.Cut(w, "=")
		if !ok || k == "" {
			continue
		}
		shouty := true
		for _, c := range k {
			if (c < 'A' || c > 'Z') && c != '_' && (c < '0' || c > '9') {
				shouty = false
				break
			}
		}
		if shouty {
			out = append(out, k)
		}
	}
	return out
}

// trimProse cuts a mined hint back to the part that is plausibly an argument.
//
// Descriptions run on: "-- infra:deploy:plan`.", "NAME=backend (the BRANCH survives)".
// The parenthetical and the trailing punctuation are commentary, not arguments.
func trimProse(rest string) string {
	end := len(rest)
	for _, stop := range []string{")", " (", ";", ", "} {
		if at := strings.Index(rest, stop); at >= 0 && at < end {
			end = at
		}
	}
	return strings.TrimRight(strings.TrimSpace(rest[:end]), ".,` ")
}

// JoinArgs is SplitArgs backwards: the line that splits back into these arguments.
//
// Needed the moment a stored argument list is put back in front of you. `-- "My Post
// Title"` comes out of the archive as two arguments, and joining them with a space would
// hand back a line that runs as four.
func JoinArgs(args []string) string {
	parts := make([]string, 0, len(args))
	for _, arg := range args {
		parts = append(parts, quoteArg(arg))
	}
	return strings.Join(parts, " ")
}

func quoteArg(arg string) string {
	if arg == "" {
		return `""`
	}
	if !strings.ContainsAny(arg, " \t'\"\\") {
		return arg
	}
	var b strings.Builder
	b.WriteByte('"')
	for _, c := range arg {
		if c == '"' || c == '\\' {
			b.WriteByte('\\')
		}
		b.WriteRune(c)
	}
	b.WriteByte('"')
	return b.String()
}

// SplitArgs splits an args line the way a shell would, so quoted arguments survive.
//
// `site:new -- "My Post Title"` has to reach go-task as one argument, not three.
func SplitArgs(input string) []string {
	out := []string{}
	var current strings.Builder
	var quote rune
	started := false

	runes := []rune(input)
	for i := 0; i < len(runes); i++ {
		c := runes[i]
		switch {
		case c == '\\':
			if i+1 < len(runes) {
				i++
				current.WriteRune(runes[i])
				started = true
			}
		case c == '\'' || c == '"':
			switch {
			case quote == c:
				quote = 0
			case quote != 0:
				current.WriteRune(c)
				started = true
			default:
				// An empty quoted string is still an argument.
				quote = c
				started = true
			}
		case unicode.IsSpace(c) && quote == 0:
			if started {
				out = append(out, current.String())
				current.Reset()
				started = false
			}
		default:
			current.WriteRune(c)
			started = true
		}
	}
	if started {
		out = append(out, current.String())
	}
	return out
}

// DangerFile is the opt-in file listing tasks that must not be run by accident.
const DangerFile = ".taskui-danger"

// GlobMatch matches a task name against a pattern supporting `*`.
//
// Deliberately not a full glob: task names are colon paths, and `deploy:*` plus exact
// names covers every real case without pulling in a dependency whose semantics would then
// need explaining.
func GlobMatch(pattern, name string) bool {
	parts := strings.Split(pattern, "*")
	first := parts[0]
	if !strings.HasPrefix(name, first) {
		return false
	}
	rest := name[len(first):]
	last := ""
	haveLast := false
	for _, part := range parts[1:] {
		last, haveLast = part, true
		if part == "" {
			continue
		}
		at := strings.Index(rest, part)
		if at < 0 {
			return false
		}
		rest = rest[at+len(part):]
	}
	// A trailing `*` swallows whatever is left; otherwise the pattern must reach the end.
	switch {
	case !haveLast:
		return rest == ""
	case last == "":
		return true
	default:
		return strings.HasSuffix(name, last)
	}
}

// DangerPatterns reads `.taskui-danger`: one pattern per line, `#` comments, blanks
// ignored.
//
// Its presence switches off the description heuristic entirely. A guess and a declaration
// disagreeing about which tasks are dangerous is worse than either alone — once you have
// written the list down, that list is the answer.
func DangerPatterns(dir string) []string {
	data, err := os.ReadFile(filepath.Join(dir, DangerFile))
	if err != nil {
		return nil
	}
	var out []string
	for line := range strings.SplitSeq(string(data), "\n") {
		line = strings.TrimSpace(strings.SplitN(line, "#", 2)[0])
		if line != "" {
			out = append(out, line)
		}
	}
	return out
}

// dangerWords mark a task as one you should not fire off from a fuzzy filter by accident.
//
// This is a heuristic over descriptions and it will both over- and under-match. It is a
// stopgap: the real fix is an explicit marker in the Taskfile, which we can honour later
// without changing anything else here.
var dangerWords = [...]string{
	"production",
	"prod database",
	"applies!",
	"wipe",
	"destroy",
	"claims ",
}

func looksDangerous(name, desc string) bool {
	d := strings.ToLower(desc)
	for _, w := range dangerWords {
		if strings.Contains(d, w) {
			return true
		}
	}
	// `backend:migrate:prod`, `backend:promo:prod`, `deploy:*`
	return strings.HasSuffix(name, ":prod") || strings.HasPrefix(name, "deploy:") || name == "deploy"
}

// parseEntry reads one line of `task --list-all`:
//
//   - build:      Build all components                    (aliases: b)
func parseEntry(line string) (Task, bool) {
	// Stripped even though Ask turns the colour off, because those are two different kinds
	// of promise: NO_COLOR is a request to a program we do not control, and what it costs
	// when it is not honoured is not an error but a project that appears to have no tasks in
	// it. This is the half we can guarantee.
	rest, ok := strings.CutPrefix(ansi.Strip(line), "* ")
	if !ok {
		return Task{}, false
	}
	// Names contain colons, so split at the first whitespace rather than the first colon —
	// `backend:migrate:down:` is one name, not three.
	name, tail := rest, ""
	if at := strings.IndexFunc(rest, unicode.IsSpace); at >= 0 {
		name, tail = rest[:at], rest[at:]
	}
	name = strings.TrimRight(name, ":")
	if name == "" {
		return Task{}, false
	}

	desc := strings.TrimSpace(tail)
	var aliases []string
	// Parsed off the end rather than searched for, so a description that happens to
	// mention the word does not get eaten.
	if strings.HasSuffix(desc, ")") {
		if at := strings.LastIndex(desc, "(aliases: "); at >= 0 {
			for a := range strings.SplitSeq(desc[at+len("(aliases: "):len(desc)-1], ",") {
				if a = strings.TrimSpace(a); a != "" {
					aliases = append(aliases, a)
				}
			}
			desc = strings.TrimSpace(desc[:at])
		}
	}

	return Task{
		Name:      name,
		Desc:      desc,
		Aliases:   aliases,
		Dangerous: looksDangerous(name, desc),
	}, true
}

// Ask builds the command for a question taskui puts to go-task, with colour off.
//
// Every question here is parsed, and go-task colours its answers whenever the environment
// tells it to — which on a CI runner is the normal case, not an exotic one. A listing line
// that arrives as "\x1b[33m* \x1b[0m\x1b[32mbuild\x1b[0m:" starts with an escape rather than
// with `* `, so the parser below matches nothing and discovery reports a project with no
// tasks in it. No error, no empty output, nothing to notice: the worst shape a bug can take.
//
// `NO_COLOR` is the standard way to say so and go-task honours it over anything asking for
// colour the other way. Deliberately not applied to the command that *runs* your task: that
// output is yours, it is shown and archived with its escapes intact, and stripping the colour
// out of a test runner's output would be taking something away rather than asking a question
// plainly.
func Ask(dir string, args ...string) *exec.Cmd {
	cmd := exec.Command("task", args...)
	cmd.Dir = dir
	cmd.Env = append(os.Environ(), "NO_COLOR=1")
	return cmd
}

// Discover runs `task --list-all` in dir and returns the tasks, minus the `*:default`
// entries — in a UI where a namespace is itself a selectable row, a task whose only job is
// "show available tasks" is noise.
func Discover(dir string) ([]Task, error) {
	declared := DangerPatterns(dir)

	cmd := Ask(dir, "--list-all")
	var stderr strings.Builder
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()
	if err != nil {
		exitError := &exec.ExitError{}
		if errors.As(err, &exitError) {
			return nil, fmt.Errorf("`task --list-all` failed in %s: %s", dir, strings.TrimSpace(stderr.String()))
		}
		return nil, fmt.Errorf("failed to run `task` — is go-task installed and on PATH? (%w)", err)
	}

	var tasks []Task
	for line := range strings.SplitSeq(string(stdout), "\n") {
		t, ok := parseEntry(line)
		if !ok || t.Name == "default" || strings.HasSuffix(t.Name, ":default") {
			continue
		}
		if len(declared) > 0 {
			t.Dangerous = false
			for _, p := range declared {
				if GlobMatch(p, t.Name) {
					t.Dangerous = true
					break
				}
			}
		}
		tasks = append(tasks, t)
	}

	sort.Slice(tasks, func(i, j int) bool { return tasks[i].Name < tasks[j].Name })
	return tasks, nil
}

// --- the JSON half -------------------------------------------------------------------

// Where a task is defined.
type Where struct {
	File string
	Line int
}

// Ok reports whether this location was actually filled in.
func (w Where) Ok() bool { return w.File != "" && w.Line > 0 }

// Detail is what `--list-all --json` knows that the text form does not: where each task is
// written, and whether go-task currently considers it up to date.
//
// Kept separate from Task because it arrives later. Discover has to return before the first
// frame; this can take four seconds on a workspace with a lot of `sources:` globs, and the
// UI is perfectly usable in the meantime without it.
type Detail struct {
	Where    Where
	UpToDate bool
}

// jsonListing is the shape of `task --list-all --json`, narrowed to what is read.
type jsonListing struct {
	Tasks []struct {
		Name     string `json:"name"`
		UpToDate bool   `json:"up_to_date"`
		Location struct {
			Taskfile string `json:"taskfile"`
			Line     int    `json:"line"`
		} `json:"location"`
	} `json:"tasks"`
}

// Details runs `task --list-all --json` and returns what it knows, keyed by task name.
//
// Slow on purpose-built repositories: computing `up_to_date` means fingerprinting every
// `sources:` glob, which on a workspace with a build directory in one means walking it.
// Measured at fifty-six times the text form on such a repo — 4.28s against 0.076s — almost
// all of it system time. Which is why nothing calls this before the first frame.
func Details(dir string) (map[string]Detail, error) {
	cmd := Ask(dir, "--list-all", "--json")
	var stderr strings.Builder
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("`task --list-all --json` failed in %s: %s", dir, strings.TrimSpace(stderr.String()))
	}

	var listing jsonListing
	if err := json.Unmarshal(stdout, &listing); err != nil {
		return nil, fmt.Errorf("could not read `task --list-all --json`: %w", err)
	}

	out := make(map[string]Detail, len(listing.Tasks))
	for _, t := range listing.Tasks {
		out[t.Name] = Detail{
			Where:    Where{File: t.Location.Taskfile, Line: t.Location.Line},
			UpToDate: t.UpToDate,
		}
	}
	return out, nil
}
